feat(gux-month-picker): created month picker component

daragh-king-genesys opened the pull request.

katie-bobbe-genesys commented:
This is looking good so far
A few observations:

- Moving focus out of the component should close the month dropdown

Accessibility:
- When focused on the December button, after pressing the right or down arrows, I think I would expect the year value to advance (this is not critical to fix)
- If you use the keyboard to change year using the `<` and `>` buttons, then press tab, the focus skips the month buttons. The month buttons seem to only be focusable after opening the month picker dropdown.

Localization related issues:

- The localization for `th` (Thai) seems to affect the year number value as well. The number in the input area does not match the year in the date picker.
![image](https://user-images.githubusercontent.com/82100934/186737507-287916fc-fa01-4ee8-8018-b71eb062233e.png)

-  For`ar` (Arabic) and `fi` (Finnish) the abbreviated month strings are too long 
![image](https://user-images.githubusercontent.com/82100934/186741652-0e2bad94-ee21-4921-99d5-cb94a867d9a7.png)

7. I think it will be odd for some languages to have the month first, then year. Usually it would be reversed in certain languages, like Chinese or Japanese. I think it would be considered to be a bug in those regions
![image](https://user-images.githubusercontent.com/82100934/186740767-9667cef1-4fbf-485d-a881-6a7cea5d8b3c.png)

MattCheely commented:
> I think it will be odd for some languages to have the month first, then year. Usually it would be reversed in certain languages, like Chinese or Japanese. I think it would be considered to be a bug in those regions

For this one, I think we should reach out to localization, but not block the PR on it, since even if the order is unusual, I don't think it will be ambiguous. If we end up merging before they get back to us, we can create a ticket to follow up.

daragh-king-genesys commented:
@katie-bobbe-genesys

1. dropdown now closes on blur
2. I am going to ask for a review but the Design System team after this is merged for feedback on the user interactions so I am not going to change when the dropdown opens or what happens on arrow presses yet. Just to note that I have confirmed that  picker and calendar are fully wired together so changing when the calendar opens will be a trivial change.
3. I added a focus trap to the popup. I will get the Design team to review this too.
4. I fixed all the localisation issues you found and will reach out to the localisation team for a full review after this PR merges.

katie-bobbe-genesys reviewed (commented):
Closing the dropdown on `blur` does not seem to be working correctly in Firefox and Safari. In Safari, clicking a month closes the dropdown without selecting the month. In Firefox, hovering the selected month closes the dropdown.

daragh-king-genesys commented on src/components/beta/gux-month-picker/gux-month-picker.tsx:
```diff
@@ -0,0 +1,345 @@
+          )
+        }
+        ref={(el: HTMLElement) => (this.monthSpinnerElement = el)}
+        aria-valuenow={this.getSpinnerValueNow('month')}
```
@katie-bobbe-genesys I don't understand what you are asking for here.

katie-bobbe-genesys replied:
Sorry I should have commented on the line below, line 238. `getSpinnerValueText()` should return `year, month` when `YearFirstLocales` is true

gavin-everett-genesys commented on src/components/beta/gux-month-picker/gux-month-calendar/gux-month-calendar.tsx:
```diff
@@ -0,0 +1,279 @@
+  }
+
+  private getYearLabel(year: string): string {
+    return new Date(Number(year), 5).toLocaleDateString(this.locale, {
+      year: 'numeric'
+    });
```
Why is '5' used here should it be '0' or null ?

gavin-everett-genesys commented on src/components/beta/gux-month-picker/gux-month-picker.tsx:
```diff
@@ -0,0 +1,353 @@
+
+import translationResources from './i18n/en.json';
+
+const DateTimeFormats = { year: 'numeric', month: 'long' };
```
Should it be called 'DateFormats' instead as there is no time involved.

gavin-everett-genesys commented:
LGTM here are a few things I spotted,

1. You can currently open the `month-picker` via mouse click but should it open as-well for keyboard (if you were to click enter on the year or month)
2. This is minor but when reading out through voice over if you hover on a month it will read the months alias eg 'APR' I am not sure if you can hide from the api but other than that accessibility is good 👍 
3. Should you be able to change the `year` by typing? The figma suggests that only the month cannot be changed via typing.
4. Should the active and hover states be the same color according to figma `@gux-blue-60`.
5. This is minor but the gap between the month-picker and month-calendar is 2px I think.
6. On the example page the title should be '`gux-month-picker-beta`'.
7. There is a small css bug on safari. The year/month seem slightly out of position. See image below, 
<img width="431" alt="safari_monthpicker" src="https://user-images.githubusercontent.com/101644078/188458689-50d28764-0c50-4b95-b48e-84154fbadb4a.png">

katie-bobbe-genesys commented on src/components/beta/gux-month-picker/gux-month-picker.tsx:
```diff
@@ -0,0 +1,353 @@
+  }
+
+  private getSpinnerValueNow(period: 'month' | 'year'): string | boolean {
+    return this.value ? getYearMonthObject(this.value)[period] : false;
```
There is an Axe accessibility violation in the "unset" month pickers: "Required ARIA attribute not present: aria-valuenow"
I don't think this is technically an accessibility violation, because `aria-valuenow` can be omitted when the value is indeterminate. I found a [related GitHub issue](https://github.com/dequelabs/axe-core/issues/2639) for the `axe-core` project, and from their answer, I think a good workaround would be to set the `aria-valuenow` to a default value.

We should also add an "unset" month picker to the tests so that the Axe scan will pick up any regressions.

daragh-king-genesys replied:
Done

daragh-king-genesys commented:
> LGTM here are a few things I spotted,
> 
> 1. You can currently open the `month-picker` via mouse click but should it open as-well for keyboard (if you were to click enter on the year or month)
> 2. This is minor but when reading out through voice over if you hover on a month it will read the months alias eg 'APR' I am not sure if you can hide from the api but other than that accessibility is good 👍
> 3. Should you be able to change the `year` by typing? The figma suggests that only the month cannot be changed via typing.
> 4. Should the active and hover states be the same color according to figma `@gux-blue-60`.
> 5. This is minor but the gap between the month-picker and month-calendar is 2px I think.
> 6. On the example page the title should be '`gux-month-picker-beta`'.
> 7. There is a small css bug on safari. The year/month seem slightly out of position. See image below,
> 

1. Leaving this for now. Will do a full review of interactivity with UX once this is merged.
2. The get an aria-label with the full month name. I suspect this is a screen reader issue not an accessibility one.
3. Yes. Joy had implemented tat functionality before I took over the component.
4. Yes.
5. Will look into this separately as it appears to be a gux-popup issue.
6. Done.
7. Will address separately.

gavin-everett-genesys reviewed (approved):
I dont know if this is an issue but I said I would raise it anyway. If you have min/max set on the calendar and dynamically set the value or hardcode it will set it to that month/year if it is out of bounds.

Example : 
```
<gux-month-picker-beta
    value="2020-01"
    min="2021-03"
    max="2023-09"
  ></gux-month-picker-beta>
```
Result : 
<img width="336" alt="min_max_calendar" src="https://user-images.githubusercontent.com/101644078/189612564-318e90fc-e878-46a6-9e28-7a6d0e0f35e7.png">

joyshaffer commented:
**UX Review:**

* Blockers to merging the PR:
    * Remove the “margin-top: 4px” on gux-month-calendar. I think this will more closely match what the other dropdown components look like in regards to space between the dropdown and dropdown list. We will be doing a visual audit in the near future, and if the spacing is to be increased, we will outline that at once for all effected components. 
    * Month button color should be gux-blue-60 in selected, active, and hover states.

* Improvements that can be made after merging:
    * Keyboard accessibility:
        * Calendar can open on enter when focused on month or year “inputs”.
        * When focused in the year header section, there seems to be a third element (I think the scrollbar) that is getting focused when I hit tab to go between the arrow buttons.
        * This also produces a kind of strange behavior to get to the month buttons because I have to tab to the page’s scrollbar and then hit the arrow key to get to the months.
        * The behavior I would expect is this:
            * I click enter on the calendar icon button to open the calendar dropdown.
            * The current selected month is focused.
            * I can use the arrow keys to navigate months.
            * I press Shift + Tab to go to the right year arrow.
            * Hitting Shift + Tab again will take me to the left year arrow. 
            * Hitting Shift + Tab again will take me to the calendar icon button, then the year, then the month input fields.
            * If I was on the left year arrow and hit Tab, it would take me to the right arrow. 
            * Hitting Tab again would focus me into January.
            * Then I could again navigate months with the arrow keys.
            * Then pressing Enter on a month I want to select or pressing Esc would close the dropdown content.
    * Gavin already notes this, but the month and year is out of alignment in Safari.

* Other notes:
    * You should not be able to change the year by typing. This is to keep the month and year “inputs” the same behavior. If a request comes in in the future for the year to be typed, then we can address that then.

Let me know if there are any concerns or questions around my comments!

daragh-king-genesys commented:
@joyshaffer All but the internal tab behaviour implemented. I will create an issue for that after this PR is merged.
